### PR TITLE
lib/promscrape: add Prometheus-compatible service discovery for Nomad

### PR DESCRIPTION
**Pull request by mr-karan**
Add `nomad_sd_configs` support for service discovery.

Closes https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3367

Here's an example file:

```
global:
  scrape_interval: 10s

scrape_configs:
  - job_name: "prom_nomad"
    nomad_sd_configs:
      - server: "localhost:4646"
        tags:
          - metrics
```

This is the result it produces:

image

The labels are similar to what is exported here: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#nomad_sd_config

### CODE REVIEW

File: lib/promscrape/discovery/nomad/api.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	return a.Config.Datacenter, nil
+}
+
+// maxWaitTime is duration for Nomad blocking request.
+func maxWaitTime() time.Duration {
+	d := discoveryutils.BlockingClientReadTimeout
+	// Nomad adds random delay up to wait/16, so reduce the timeout in order to keep it below BlockingClientReadTimeout.
+	// See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
+	// The timeout cannot exceed 10 minuntes. See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/api.go:136]:
```suggestion
	// Nomad adds random delay up to wait/16, so reduce the timeout in order to keep it below BlockingClientReadTimeout.
	// See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
	d -= d / 16
	// The timeout cannot exceed 10 minuntes. See https://developer.hashicorp.com/nomad/api-docs#blocking-queries
	
```

File: lib/promscrape/discovery/nomad/api.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	return cfg, nil
+}
+
+func getToken(token *promauth.Secret) (string, error) {
+	if token != nil {
+		return token.String(), nil
+	}
+	t := os.Getenv("NOMAD_TOKEN")
+	// Allow empty token - it shouls work if ACL is disabled in Nomad.
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/api.go:111]:
```suggestion
	// Allow empty token - it should work if ACL is disabled in Nomad.
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+type serviceWatcher struct {
+	serviceName string
+	services    []Service
+	stopCh      chan struct{}
+}
+
+// newnomadWatcher creates new watcher and starts background service discovery for Consul.
+func newnomadWatcher(client *discoveryutils.Client, sdc *SDConfig, datacenter, namespace string) *nomadWatcher {
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:45]:
```suggestion
func newNomadWatcher(client *discoveryutils.Client, sdc *SDConfig, datacenter, namespace string) *nomadWatcher {
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+	cw.servicesLock.Lock()
+	sns := make(map[string][]Service, len(cw.services))
+	for svc, sw := range cw.services {
+		sns[svc] = sw.services
+	}
+	cw.servicesLock.Unlock()
+	return sns
+}
+
+// watchForServiceNodesUpdates watches for Consul serviceNode changes for the given serviceName.
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:224]:
```suggestion
// watchForServiceNodesUpdates watches for Nomad serviceNode changes for the given serviceName.
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+// watchForServiceNodesUpdates calls initWG.Done() once the initialization is complete and the first discovery iteration is done.
+func (sw *serviceWatcher) watchForServiceAddressUpdates(nw *nomadWatcher, initWG *sync.WaitGroup) {
+	clientAddr := nw.client.Addr()
+	index := int64(0)
+	// TODO: Maybe use a different query arg.
+	path := "/v1/service/" + sw.serviceName + nw.serviceNamesQueryArgs
+	f := func() {
+		data, newIndex, err := getBlockingAPIResponse(nw.client, path, index)
+		if err != nil {
+			logger.Errorf("cannot obtain Consul serviceNodes for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:235]:
```suggestion
			logger.Errorf("cannot obtain Nomad services for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+			return
+		}
+		if index == newIndex {
+			// Nothing changed.
+			return
+		}
+		sns, err := parseServices(data)
+		if err != nil {
+			logger.Errorf("cannot parse Consul serviceNodes response for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:244]:
```suggestion
			logger.Errorf("cannot parse Nomad services response for serviceName=%q from %q: %s", sw.serviceName, clientAddr, err)
```

File: lib/promscrape/discovery/nomad/service.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discoveryutils"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promutils"
+)
+
+// getServiceLables returns labels for Nomad services with given cfg.
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/service.go:13]:
```suggestion
// getServiceLabels returns labels for Nomad services with given cfg.
```

File: lib/promscrape/discovery/nomad/service.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	AllocID     string   `json:"AllocID"`
+	Tags        []string `json:"Tags"`
+	Address     string   `json:"Address"`
+	Port        int      `json:"Port"`
+}
+
+func parseServices(data []byte) ([]Service, error) {
+	var sns []Service
+	if err := json.Unmarshal(data, &sns); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal ServiceNodes from %q: %w", data, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/service.go:51]:
```suggestion
		return nil, fmt.Errorf("cannot unmarshal Services from %q: %w", data, err)
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+	stopCh chan struct{}
+}
+
+type serviceWatcher struct {
+	serviceName string
+	services    []Service
+	stopCh      chan struct{}
+}
+
+// newnomadWatcher creates new watcher and starts background service discovery for Consul.
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:44]:
```suggestion
// newNomadWatcher creates new watcher and starts background service discovery for Nomad.
```

File: lib/promscrape/discovery/nomad/watch.go [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+	for serviceName, sw := range cw.services {
+		if _, ok := newServiceNamesMap[serviceName]; ok {
+			continue
+		}
+		close(sw.stopCh)
+		delete(cw.services, serviceName)
+		serviceWatchersStopped.Inc()
+
+		// Do not wait for the watcher goroutine to exit, since this may take for up to maxWaitTime
+		// if it is blocked in Consul API request.
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/watch.go:116]:
```suggestion
		// if it is blocked in Nomad API request.
```

File: lib/promscrape/discovery/nomad/api.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+		if len(ind) == 0 {
+			logger.Errorf("cannot find X-Nomad-Index header in response from %q", path)
+			return
+		}
+		newIndex, err := strconv.ParseInt(string(ind), 10, 64)
+		if err != nil {
+			logger.Errorf("cannot parse X-Nomad-Index header value in response from %q: %s", path, err)
+			return
+		}
+		// Properly handle the returned newIndex according to https://www.consul.io/api-docs/features/blocking#implementation-details
```
Comment (hagen1778) [inline lib/promscrape/discovery/nomad/api.go:162]: Is the index logic the same for Nomad as for Consul?

**Review by hagen1778 — COMMENTED** [review]
LGTM! Please see my comments

@f41gh7 @zekker6 wdyt?

File: lib/promscrape/discovery/nomad/api.go (hunk 11 above)
Comment (zekker6) [reply lib/promscrape/discovery/nomad/api.go:162]: Yes, according to https://developer.hashicorp.com/nomad/api-docs#blocking-queries

**Review by zekker6 — APPROVED** [review]
LGTM

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by mr-karan** [comment]
@hagen1778 Thanks a lot for the review. I've pushed the changes.

**Review by hagen1778 — COMMENTED** [review]
Would also be great to add a note to the docs/CHANGELOG.md

**Comment by hagen1778** [comment]
@mr-karan please check the CI build issues. Thanks!

**Comment by mr-karan** [comment]
@hagen1778 Added a comment clarifying that indexing logic is same. Nomad docs for indexing behaviour don't exist as of now. Also fixed the build issues and added a small note in docs/CHANGELOG.md

**Comment by hagen1778** [comment]
Looks very good! Thanks for your contribution!

**Pull request merged**